File: pt-2/481.magical-string.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
public:
    int magicalString(int n) {
        if (n <= 0) return 0;
        vector<int> s = {1, 2, 2};
        if (n <= 3) return count(s.begin(), s.begin() + n, 1);
        int read = 2, nextNum = 1, ones = 1;
        while ((int)s.size() < n) {
            int repeat = s[read++];
            for (int k = 0; k < repeat && (int)s.size() < n; ++k) {
                s.push_back(nextNum);
                if (nextNum == 1) ++ones;
            }
            nextNum = 3 - nextNum;
        }
        return ones;
    }
};

/*
Interview explanation:
The magical string describes its own run lengths. Starting from 122, a read pointer gives how many copies of the next alternating value to append.

C++ data structures: vector<int> stores generated digits because future positions are read as run lengths.

Edge cases: n<=3 is answered directly from the seed.

Complexity: O(n) time and O(n) space.
*/

File: pt-2/481.magical-string.go
package leetcode

// MagicalString481 generates the self-describing run-length string from the
// seed 122. The read pointer supplies the next run length; nextValue alternates
// between 1 and 2. Ones are counted only until length n is reached.
//
// Go data structure note: []int is a growable slice because generated values
// become future run lengths.
//
// Time: O(n). Space: O(n).
func MagicalString481(n int) int {
	if n <= 0 {
		return 0
	}
	s := []int{1, 2, 2}
	if n <= 3 {
		ans := 0
		for i := 0; i < n; i++ {
			if s[i] == 1 {
				ans++
			}
		}
		return ans
	}
	read, nextNum, ones := 2, 1, 1
	for len(s) < n {
		repeat := s[read]
		for i := 0; i < repeat && len(s) < n; i++ {
			s = append(s, nextNum)
			if nextNum == 1 {
				ones++
			}
		}
		nextNum = 3 - nextNum
		read++
	}
	return ones
}
